cmd/erasure-server-pool-decom.go

```diff
@@ -0,0 +1,909 @@
+
+	if idx+1 > len(z.poolMeta.Pools) {
+		return PoolStatus{}, errInvalidArgument
+	}
+
+	pool := z.serverPools[idx]
+	info, _ := pool.StorageInfo(ctx)
+	info.Backend.Type = madmin.Erasure
+
+	scParity := globalStorageClass.GetParityForSC(storageclass.STANDARD)
+	if scParity <= 0 {
+		scParity = z.serverPools[0].defaultParityCount
+	}
+	rrSCParity := globalStorageClass.GetParityForSC(storageclass.RRS)
+	info.Backend.StandardSCData = append(info.Backend.StandardSCData, pool.SetDriveCount()-scParity)
+	info.Backend.RRSCData = append(info.Backend.RRSCData, pool.SetDriveCount()-rrSCParity)
+	info.Backend.StandardSCParity = scParity
+	info.Backend.RRSCParity = rrSCParity
+
+	currentSize := TotalUsableCapacityFree(info)
+
+	poolInfo := z.poolMeta.Pools[idx]
+	if poolInfo.Decommission != nil {
+		poolInfo.Decommission.CurrentSize = currentSize
+	} else {
+		poolInfo.Decommission = &PoolDecommissionInfo{
+			CurrentSize: currentSize,
+			TotalSize:   TotalUsableCapacity(info),
+		}
+	}
+	return poolInfo, nil
+}
+
+func (z *erasureServerPools) ReloadPoolMeta(ctx context.Context) (err error) {
+	meta := poolMeta{}
+
+	if _, err = meta.load(ctx, z.serverPools[0], z.serverPools); err != nil {
+		return err
+	}
+
+	z.poolMetaMutex.Lock()
+	defer z.poolMetaMutex.Unlock()
+
+	z.poolMeta = meta
+	return nil
+}
+
+func (z *erasureServerPools) DecommissionCancel(ctx context.Context, idx int) (err error) {
+	if idx < 0 {
+		return errInvalidArgument
+	}
+
+	if z.SinglePool() {
+		return errInvalidArgument
+	}
+
+	z.poolMetaMutex.Lock()
+	defer z.poolMetaMutex.Unlock()
+
+	meta := z.poolMeta
+	if meta.DecommissionCancel(idx) {
+		z.decommissionCancelers[idx]() // cancel any active thread.
+		defer func() {
+			if err == nil {
+				z.poolMeta.DecommissionCancel(idx)
+				globalNotificationSys.ReloadPoolMeta(ctx)
+			}
+		}()
+		return meta.save(ctx, z.serverPools)
+	}
+	return nil
+}
+
+func (z *erasureServerPools) DecommissionFailed(ctx context.Context, idx int) (err error) {
+	if idx < 0 {
+		return errInvalidArgument
+	}
+
+	if z.SinglePool() {
+		return errInvalidArgument
+	}
+
+	z.poolMetaMutex.Lock()
+	defer z.poolMetaMutex.Unlock()
+
+	meta := z.poolMeta
+	if meta.DecommissionFailed(idx) {
+		z.decommissionCancelers[idx]() // cancel any active thread.
+		defer func() {
+			if err == nil {
+				z.poolMeta.DecommissionFailed(idx)
+				globalNotificationSys.ReloadPoolMeta(ctx)
+			}
+		}()
+		return meta.save(ctx, z.serverPools)
+	}
+	return nil
+}
+
+func (z *erasureServerPools) CompleteDecommission(ctx context.Context, idx int) (err error) {
+	if idx < 0 {
+		return errInvalidArgument
+	}
+
+	if z.SinglePool() {
+		return errInvalidArgument
+	}
+
+	z.poolMetaMutex.Lock()
+	defer z.poolMetaMutex.Unlock()
+
+	meta := z.poolMeta
+	if meta.DecommissionComplete(idx) {
+		defer func() {
+			if err == nil {
+				z.poolMeta.DecommissionComplete(idx)
+				globalNotificationSys.ReloadPoolMeta(ctx)
+			}
+		}()
+		return meta.save(ctx, z.serverPools)
+	}
+	return nil
```

Comment:
We could do something like this to avoid calling `DecommissionComplete` twice.
```suggestion
	meta := z.poolMeta
	if meta.DecommissionComplete(idx) {
	    if err = meta.save(ctx, z.serverPools); err == nil {
	        globalNotificationSys.ReloadPoolMeta(ctx)
	    }
	}
	return err
```